src/workloads/execution/Densify.yml

```diff
@@ -0,0 +1,421 @@
+SchemaVersion: 2018-07-01
+Owner: "@mongodb/query"
+Description: |
+  This workload test tests the performance of the $densify stage. $densify can operate on both dates
+  and numbers.
+
+Actors:
+- Name: InsertData
+  Type: Loader
+  Threads: 1
+  Phases:
+  - Repeat: 1
+    Database: &db test
+    Threads: 1
+    CollectionCount: 1
+    DocumentCount: 10000
+    BatchSize: &batchSize 1000
+    Document:
+      partitionKey: {^RandomInt: {min: 1, max: 100}}
+      # Test a subset of units that have different characteristics.
+      number: {^RandomInt: {min: 0, max: 20000}} # Seperate codepath for numeric DensifyValues.
+      milliseconds: {^RandomDate: {min: "2021-01-01T00:00:00.000Z", max: "2021-01-01T00:00:20.000Z"}} # Most basic date unit.
+      hours: {^RandomDate: {min: "2021-01-01", max: "2023-01-01"}} # Date unit with constant number of milliseconds.
+      months: {^RandomDate: {min: "2021-01-01", max: "4000-01-01"}} # Date unit with non-constant number of milliseconds.
+  - &Nop {Nop: true}
+  - *Nop
+  - *Nop
+
+# $densify stages always add a $sort on the field that is being densified, so this phase adds 
+# indexes on the date/number fields so that we aren't performing an in-memory sort.
+- Name: CreateIndex
+  Type: RunCommand
+  Threads: 1
+  Phases:
+  - *Nop
+  - Repeat: 1
+    Database: *db
+    Operations:
+    - OperationName: RunCommand
+      OperationCommand:
+        createIndexes: Collection0
+        indexes: 
+        - key: {number: 1}
+          name: number
+        - key: {milliseconds: 1}
+          name: milliseconds
+        - key: {hours: 1}
+          name: hours
+        - key: {months: 1}
+          name: months
+  - *Nop
+  - *Nop
+
+- Name: Quiesce
+  Type: QuiesceActor
+  Threads: 1
+  Database: *db
+  Phases:
+  - *Nop
+  - *Nop
+  - Repeat: 1
+  - *Nop
+
+- Name: Densify
+  Type: RunCommand
+  Threads: 1
+  Phases:
+  - *Nop
+  - *Nop
+  - *Nop
+  - Repeat: 10
+    Database: *db
+    Operations:
+
+    # Metrics are grouped by the step unit they operate with (numiric, millisecond, hour, month).
+    # Each group has a metric for each possible codepath (Partition/no partition, explicit/full/partitioned bounds).
+
+    # Numeric metrics. 
+    - OperationMetricsName: NumericFullDensify
+      OperationName: RunCommand
+      OperationCommand:
+        aggregate: Collection0
+        pipeline: [{
+          $densify: {
+            field: "number", 
+            range: {
+              bounds: "full", 
+              step: {^RandomInt: {min: 1, max: 11}}
```

Comment:
If I'm understanding $densify / this workload correctly - each document will have a `number` field that is a value from 0 - 20000. Since there are 10000 documents in the collection, there will be many "gaps" in between the `number` field for these documents, but there is a very small probability that there would be a "gap" that is as big as 10 documents (ie. we have a doc with `number: 0` but the second in the sorted order is `number: 11`).

For $densify does this mean that, if the step size is a certain number (let's say 10) we would only generate new documents if there is a "gap" in the data that is that big? And if so, have you done the [noise/robustness testing](https://docs.google.com/document/d/1336S0iRRqNBDc_xgnmqOFoIx6G4_8FJiSZg0vpQMRBE/edit#heading=h.bwieg7qd7dzx) for these workloads? I can imagine that if the first question is true (and let me know if I'm missing something!), then we would have a lot of variance in these results depending on the step size that is chosen on each repeat - a better approach to this kind of perf testing might be to have separate operations with different step sizes (for example NumericFullDensifySmallStepSize and NumericFullDensifyLargeStepSize).

Reply:
To my knowledge $densify doesn't look for a gap of a certain size, but rather generates documents that are on-step relative to a lower bound that's either explicitly defined, or is the minimum value in the range/partition. So a collection with the values [2,5,7,9,13,14,15,19] and a densify operation with `{range: "full", step: 4}` would output [2,5,**6**,7,9,**12**,13,14,15,**16**,19] (inserted documents in bold) even though the gaps between documents are at their smallest there.

Step sizes will have an impact on the number of documents generated, though, which could impact runtime. It might be a good idea to split up the metrics into smaller and larger steps.

I haven't done noise/robustness testing though, and that's definitely something I missed! I'll be sure to queue up more builds to double check.

Reply:
Thanks for clarifying that the step size is relative to the lower bound / min value specifically. Even with this, I agree that a widely-varying step size would have a not-insignificant impact on the latency results (the specifics of which I'm looking forward to seeing from your tests) so separate metrics would still be useful to have!

Reply:
Just split up the tests into a small step (1-3) and a large step case (7-11). It blew up the size of the file, but I still think it's pretty readable. I could very well be misguided on that since I also put the file together, so let me know on that front